### PR TITLE
MessagePack extensions for the datetime and interval types

### LINKED ISSUE #3060: MessagePack extensions for the datetime and interval types
**Related dev. issue(s):** https://github.com/tarantool/tarantool/pull/6660/commits

**Product:** Tarantool
**Since:** 2.10.0
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/
**SME:** @ igormunkin, @ tsafin  

# Details

- [x] Describe new MessagePack extensions for the datetime type (similar to the description for UUID and decimal)
- [x] Describe new MessagePack extensions for the interval type
- [x] Link with the description of types here https://www.tarantool.io/en/doc/latest/book/box/data_model/#field-type-details
- [x] Update .po files

Input from RnD:
- https://github.com/tarantool/tarantool/wiki/Datetime-Internals#messagepack-encoding-schema
- https://hackmd.io/@Mons/S1Vfc_axK#%D0%A3%D0%BF%D0%B0%D0%BA%D0%BE%D0%B2%D0%BA%D0%B0-%D0%B2-msgpack

Follow-up of https://github.com/tarantool/doc/issues/2509

### PR DESCRIPTION
**Pull request by xuniq**
Fixes #3060

### CODE REVIEW

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 1]
```diff
@@ -233,3 +234,41 @@ the server response will look like this:
...
+
+Since version :doc:`2.10.0 </release/2.10.0>`.
+``MP_DATETIME`` is 4.
+
+The datetime MessagePack representation looks like this:
+
+.. code-block:: none
+
+    +--------+-------------------+------------+===============+
+    | MP_EXT | length (optional) | MP_DECIMAL | PackedDecimal |
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: Wrong picture - it's supposed to be datetime, not decimal.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 2]
```diff
@@ -233,3 +234,41 @@ the server response will look like this:
...
+Since version :doc:`2.10.0 </release/2.10.0>`.
+
+``MP_INTERVAL`` is 6.
+
+The interval MessagePack representation looks like this:
+
+.. code-block:: none
+
+    +--------+-------------------+------------+===============+
+    | MP_EXT | length (optional) | MP_INTERVAL | PackedDecimal |
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: Again, it's depicting decimal, not interval.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 3]
```diff
@@ -245,15 +247,28 @@ The DATETIME type
...
 
+Datetime MessagePack serialization schema is MP_EXT extension, which creates container of 8 or 16 bytes long payload.
+
 The datetime MessagePack representation looks like this:
 
-.. code-block:: none
+..  code-block:: none
+
+    +---------+--------+===============+-------------------------------+
+    |0xd7/0xd8|type (4)| seconds (8b)  | nsec; tzoffset; tzindex; (8b) |
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: Oh, now it's much better, and is more relevant!

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 4]
```diff
@@ -267,8 +282,109 @@ Since version :doc:`2.10.0 </release/2.10.0>`.
...
+..  code-block:: none
 
-    +--------+-------------------+------------+===============+
+    +--------+-------------------+------------+================+
     | MP_EXT | length (optional) | MP_INTERVAL | PackedDecimal |
-    +--------+-------------------+------------+===============+
+    +--------+-------------------+------------+================+
+
+MessagePack encoding schema for intervals is a little bit more complex than one used by datetime values
+(strictly speaking - they has nothing in common whatsoever).
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: > (strictly speaking - they has nothing in common whatsoever)

I'd remove this part from official documentation.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 5]
```diff
@@ -267,8 +282,109 @@ Since version :doc:`2.10.0 </release/2.10.0>`.
...
+
+..  code-block:: none
+
+    <--------- length bytes --------->
+    +----------+=====================+
+    | field ID |     field value     |
+    +----------+=====================+
+
+Here ``field ID`` is either ``MP_INT`` or ``MP_UINT``. |br|
+``scale`` = number of digits after the decimal point
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: Which `scale`?

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 6]
```diff
@@ -267,8 +282,109 @@ Since version :doc:`2.10.0 </release/2.10.0>`.
...
+
+``BCD`` is a sequence of bytes representing decimal digits of the encoded number
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: `BCD` ? Smells like copy-paste from decimal

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 7]
```diff
@@ -245,48 +249,51 @@ The DATETIME type
...
 
-    +--------+-------------------+------------+===============+
-    | MP_EXT | length (optional) | MP_DECIMAL | PackedDecimal |
-    +--------+-------------------+------------+===============+
-
-MessagePack data encoded using fixext8 (0xd7) or fixext16 (0xd8), and may contain:
+MessagePack data contains:
 
-[required] seconds parts as full, unencoded, signed 64-bit integer, stored in little-endian order;
+*   Seconds (8b) as full, unencoded, signed 64-bit integer, stored in little-endian order.
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: `Seconds (8 bytes)`

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 8]
```diff
@@ -245,48 +249,51 @@ The DATETIME type
...
 
-[optional] all the other fields (nsec, tzoffset, tzindex) if any of them were having not 0 value.
-They are packed naturally in little-endian order;
+*   The optional fields (8b), if any of them have a non-zero value.
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: `The optional fields (also 8 bytes)...`

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 9]
```diff
@@ -323,33 +328,9 @@ List of IDs:
 *   7 -- nanosecond
 *   8 -- adjust
 
-``PackedInterval`` has the following structure:
```
Comment (tsafin) [inline doc/dev_guide/internals/msgpack_extensions.rst]: In this commit this confusing copy/paste has been deleted. Nice!

**Review by tsafin — CHANGES_REQUESTED** [review]
Sorry for the confusing references - I've commented some particular commits inside of pull-request, and sometimes those comments have already been resolved till the end. But there are some complaints which are still active.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 10]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+..  _msgpack_ext-datetime:
+
+**********************************
+The DATETIME type
+**********************************
+
+Since version :doc:`2.10.0 </release/2.10.0>`.
+The MessagePack EXT type ``MP_EXT`` together with the extension type
+``MP_DATETIME`` is a header for values of the DATETIME type.
+It creates a container of 8 or 16 bytes long payload.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: "long payload" -- is it a term, or we just mean "8 or 16 bytes long"? If the second case, please rephrase.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 11]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+
+``MP_DATETIME`` type is 4.
+
+The `MessagePack spec <https://github.com/msgpack/msgpack/blob/master/spec.md>`_
+defines ``d7``to mean ``fixext`` with size 8 or ``d8`` to mean ``fixext`` with size 16.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]:
```suggestion
defines ``d7`` to mean ``fixext`` with size 8 or ``d8`` to mean ``fixext`` with size 16.
```

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 12]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+**********************************
+
+Since version :doc:`2.10.0 </release/2.10.0>`.
+The MessagePack EXT type ``MP_EXT`` together with the extension type
+``MP_DATETIME`` is a header for values of the DATETIME type.
+It creates a container of 8 or 16 bytes long payload.
+
+``MP_DATETIME`` type is 4.
+
+The `MessagePack spec <https://github.com/msgpack/msgpack/blob/master/spec.md>`_
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]:
```suggestion
The `MessagePack specification <https://github.com/msgpack/msgpack/blob/master/spec.md>`_
```

It's better not to use slang terms.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 13]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+..  code-block:: none
+
+    +---------+----------------+==========+-----------------+
+    | MP_EXT  | MP_DATETIME    | seconds  | nsec; tzoffset; |
+    | = d7/d8 | = 4            |          | tzindex;        |
+    +---------+----------------+==========+-----------------+
+
+MessagePack data contains:
+
+*   Seconds (8b) as full, unencoded, signed 64-bit integer, stored in little-endian order.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: "as full" -- what does it mean?

Please also correct the punctuation and articles.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 14]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+
+*   The optional fields (8b), if any of them have a non-zero value.
+    The fields include nsec, tzoffset, and tzindex, packed in little-endian order.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]:
```suggestion
    The fields include ``nsec``, ``tzoffset``, and ``tzindex`` packed in the little-endian order.
```

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 15]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+
+For more information about datetime, see :ref:`datetime field type details <index-box_datetime>`
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: "For more information about datetime, ..."

There should be "the datetime type" or something similar -- there's no such word/entity as just "datetime".

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 16]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+and :doc:`module datetime </reference/reference_lua/datetime>`.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: > and :doc:`module datetime </reference/reference_lua/datetime>`

Better to say, "description of the datetime module" or "reference for the datetime module".

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 17]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+The INTERVAL type
+**********************************
+
+Since version :doc:`2.10.0 </release/2.10.0>`.
+The MessagePack EXT type ``MP_EXT`` together with the extension type
+``MP_INTERVAL`` is a header for values of the INTERVAL type.
+
+``MP_INTERVAL`` type is 6.
+
+The interval is saved as variant of a map with predefined number of known attributes names.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: Correct articles.

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 18]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+..  code-block:: none
+
+    +--------+-------------------------+-------------+----------------+
+    | MP_EXT | Size of packed interval | MP_INTERVAL | PackedInterval |
+    +--------+-------------------------+-------------+----------------+
+
+Packed interval consists of:
+
+*   packed number of non-zero fields
+*   packed non-null fields
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]:
```suggestion
*   packed non-null fields.
```

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 19]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+**********************************
+
+Since version :doc:`2.10.0 </release/2.10.0>`.
+The MessagePack EXT type ``MP_EXT`` together with the extension type
+``MP_INTERVAL`` is a header for values of the INTERVAL type.
+
+``MP_INTERVAL`` type is 6.
+
+The interval is saved as variant of a map with predefined number of known attributes names.
+Some attributes might be undefined. In this case, they will be omitted from the generated payload.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: > might be

Do we really need to use this modal verb?

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 20]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+*   3 -- day
+*   4 -- hour
+*   5 -- minute
+*   6 -- second
+*   7 -- nanosecond
+*   8 -- adjust
+
+**Example**
+
+Interval value ``1 year, 200 months, -77 days`` is encoded in the following way:
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]:
```suggestion
Interval value ``1 years, 200 months, -77 days`` is encoded in the following way:
```

File: doc/dev_guide/internals/msgpack_extensions.rst [hunk 21]
```diff
@@ -233,3 +240,141 @@ the server response will look like this:
...
+*   01 -- packed value ``1``
+*   01 -- field ID (month)
+*   CCC8 -- packed value ``200``
+*   03 -- field ID (day)
+*   D0B3 -- packed value ``-77``
+*   08 -- field ID (adjust)
+*   01 -- packed value ``1`` (DT_LIMIT)
+
+For more information about intervals, see :ref:`interval field type details <index-box_interval>`
+and :doc:`module datetime </reference/reference_lua/datetime>`.
```
Comment (veod32) [inline doc/dev_guide/internals/msgpack_extensions.rst]: The same comment as to the last sentense in the section for the DATETIME type.

**Review by veod32 — CHANGES_REQUESTED** [review]

**Review by tsafin — APPROVED** [review]
LGTM, Thanks!

**Review by veod32 — APPROVED** [review]

**Pull request merged**